File: 1208.尽可能使字符串相等.py
class Solution:
    def equalSubstring(self, s: str, t: str, maxCost: int) -> int:
        '''
        只能s的第i个变到t的第i个
        维护一个动态的滑动窗口，right端加入变化的字母，在cost大于maxcost的时候left端减少
        记录窗口的最大长度
        '''
        window_len=0
        max_window=0
        right=0
        while right<len(s) :
            if abs(ord(s[right])-ord(t[right]))<=maxCost :
                window_len=window_len+1
                maxCost=maxCost-abs(ord(s[right])-ord(t[right]))
                right=right+1
                #print(right)
            else :
                if window_len>0 :
                    maxCost=maxCost+abs(ord(s[right-window_len])-ord(t[right-window_len]))
                    window_len=window_len-1
                else :
                    right=right+1
    
            max_window=max(max_window,window_len)
        return max_window

File: 1208.尽可能使字符串相等.cpp
/*
 * @lc app=leetcode.cn id=1208 lang=cpp
 *
 * [1208] 尽可能使字符串相等
 */

// @lc code=start
class Solution {
public:
    int equalSubstring(string s, string t, int maxCost) {
        vector<int> cost(s.size(), 0);

        for (int i = 0; i < s.size();++i) {
            cost[i] = abs(int(s[i]) - int(t[i]));
        }

        int window_len = 0,max_window=0;
        int right = 0;
        while (right<s.size()) {
            if (cost[right]<=maxCost) {
                maxCost -= cost[right];
                window_len++;
                right++;
            }
            else {
                if (window_len>0) {
                    maxCost += cost[right - window_len];
                    window_len--;
                }
                else {
                    right++;
                }
            }
            max_window = max(max_window, window_len);
        }
        return max_window;
    }
};
// @lc code=end

